README.md
# OCCA-XSBench

## Introduction

OCCA-XSBench is a port of XSBench, a neutronics proxy-app. It uses the OCCA
multithreading API to target both multicore CPUs and many-core hardware
accellerators.  OCCA-XSBench features a unified, highly-optimized kernel that
may be compiled for either OpenMP and CUDA.  

XSBench and OCCA were developed by research groups affiliated with the [Center
for Exascale Simulation of Advanced Reactors](https://cesar.mcs.anl.gov/), a
DOE Office of Science Co-design Center.  

[XSBench][xsbench] is a mini-app representing a key computational kernel of the
Monte Carlo neutronics application [OpenMC][openmc].  It was originally
developed by John Tramm, Andrew Siegel, et al. in the Mathematics and Computer
Science Division at Argonne National Laboratory.  A full explanation of the
theory and purpose of XSBench is provided in `docs/XSBench_Theory.pdf`.  

[OCCA][occa] is an extensible multi-threading programming API that allows a
unified source code to target many multi-threading languages, including OpenMP,
CUDA, and OpenCL.  It was originally developed by David Medina, Tim Warburton,
et. al in the Department of Computational and Applied Mathematics at Rice
University.  Documentation and downloads for OCCA can be found at
[libocca.org][occa].

For more information about OCCA-XSBench, please contact Ron
Rahaman \(rahaman@anl.gov \) and Amanda Lund \(alund@anl.gov \) at Argonne National
Laboratory.  

[xsbench]: https://github.com/ANL-CESAR/XSBench "XSBench"
[openmc]:  https://mit-crpg.github.io/openmc/ "OpenMC"
[occa]:  http://libocca.org "OCCA"


## Installation

#### Dependencies

The OCCA library must first be installed and configured.  OCCA-XSBench
currently supports OpenMP and CUDA modes; hence, OCCA must be compiled with
OpenMP and CUDA modes enabled. 

#### Compilation

The XSBench driver may be compiled with `make`.  The makefile
(`src/Makefile`) has the following settings:

* `COMPILER = ( gnu | intel ):` Specifies compiler for driver.
* `OPTIMIZE = ( yes | no ):` Sets the -O3 flag.  
* `DEBUG  = ( yes | no ):` Sets the -g flag.
* `PROFILE = ( yes | no );` Sets the -pg flag.
* `VERIFY = ( yes | no ):` Enables verification mode.  See [Verification](#verification), below.
* `BINARY_DUMP = ( yes | no ):`  Writes cross-sections to file.  See [Binary
  dump/read](#binary dump/read), below.
* `BINARY_READ = ( yes | no ):` Reads cross-sections from file.  See [Binary
  dump/read](#binary dump/read), below.

The XSBench kernel is compiled at runtime using the environment variabls:
`$OCCA_OPENMP_COMPILER`, `$OCCA_OPENMP_COMPILER_FLAGS`, `$OCCA_CUDA_COMPILER`,
and `$OCCA_CUDA_COMPILER_FLAGS`.  

Currently, the multithreading mode \(OpenMP or CUDA\) is specified by a macro
in `src/Main.c`.  Future versions will allow the mode to be set at runtime.  

##Usage

#### Synopsis

`./XSBench [-t <threads>] [-s 'small'|'large'|'XL'|'XXL'] [-g <gridpoints>] [-l <lookups>] [-m <OCCA mode>] [-d <OCCA device ID>] [-p <OCCA platform ID>] [-o <OCCA outer dimension>] [-i <OCCA inner dimension>] [-k <OCCA kernel>]`

#### Options

- `-t <threads>`

  Sets the number of OpenMP threads to run. Defaults to the value of
  `$OMP_NUM_THREADS`, if it is set. Otherwise, defaults to
  `omp_get_num_procs()`.
 

- `-s 'small'|'large'|'XL'|'XXL'`

  Sets the size of the Hoogenboom-Martin reactor model. Defaults to `'large'`.
  
  The H-M size corresponds to the number of nuclides present
  in the fuel region.  The `'small'` version has 34 fuel nuclides,
  whereas the `'large'` version has 321 fuel nuclides. This
  significantly slows down the runtime of the program as the
  data structures are much larger, and more lookups are required
  whenever a lookup occurs in a fuel material.    

  The additional size options, `'XL'` and `'XXL'`, do not directly correspond
  to any particular physical model. They are similar to the H-M "large" option,
  except the number of gridpoints per nuclide has been increased greatly. This
  creates an extremely large energy grid data structure \(XL: 120GB, XXL:
  252GB\), which is unlikely to fit on a single node, but is useful for
  experimentation purposes on novel architectures.

- `-g <gridpoints>`

  Sets the number of gridpoints per nuclide. Defaults to 11,303. 
  
  This corresponds to the average number of actual gridpoints per nuclide in
  the H-M Large model as run by OpenMC with the actual ACE ENDF cross-section
  data. 

- `-l <lookups>`

  Sets the number of cross-section \(XS\) lookups to perform. Defaults to
  15,000,000. 
  
  Users may want to increase this value if they wish to extend the runtime of
  XSBench, perhaps to produce more reliable performance counter data - as
  extending the run will decrease the percentage of runtime spent on
  initialization.

- `-m <OCCA mode>`

  Sets the OCCA compilation mode (OpenMP, CUDA, OpenCL, etc.).  Defaults to OpenMP.

- `-d <OCCA device ID>`

  Sets the ID for the device used by OCCA.  Defaults to 0.  

- `-p <OCCA platform ID>`

  Sets the ID for the platform used by OCCA.  Defaults to 0.  

- `-o <OCCA outer dimension>`

  Size of the OCCA outer dimension. Defaults to 468750.

- `-i <OCCA inner dimension>`

  Size of the OCCA inner dimension.  Defaults to 32.

- `-k <OCCA kernel>`

  The OKL source file used to compile the main lookup kernel.  Defaults to hybridLookupKernel.okl

#### Verification

  If compiled in verification mode \(see [Compilation](#compilation)\), a weak test for
  consistency is performed.  The program will accumulate cross-section values
  and report an non-deteriministic expectation value \(approximately 23.5 +/-
  0.5\ for the 'small' benchmark and 73.2 +/- 0.2 for the 'large' benchmark).  

  The verification does not produce a deterministic value because the random
  number stream is not reproducible and because the floating-point arithmatic
  is non-associative.  Future versions will produce a deterministic integer
  checksum.  

#### Binary dump/read

  If compiled in `BINARY_DUMP` mode \(see [Compilation](#compilation)\),
  XSBench writes a binary file containing a randomized data set of cross
  sections. This can be used in tandem with the binary read mode to skip
  generation of cross section data every time the program is run.

  If compiled in `BINARY_READ` mode \(see [Compilation](#compilation)\),
  XSBench reads the binary file created by the binary dump mode as a (usually)
  much faster substitution for randomly generating XS data on-the-fly. This mode
  is particularly useful if running on simulators where walltime minimization is
  extremely critical for logistical reasons.

## References

- J. R. Tramm, A. R. Siegel, T. Islam, and M. Schulz. "XSBench - The
  Development and Verification of a Performance Abstraction for Monte
  Carlo Reactor Analysis." *PHYSOR 2014 - The Role
  of Reactor Physics toward a Sustainable Future*, Kyoto, 2014. [pdf](http://www.mcs.anl.gov/papers/P5064-0114.pdf)
- D. S. Medina, M. St-Cyr, T. Warburton.  "OCCA: A unifiied approach to
  multi-threading languages." *CoRR* abs/1403.0968, 2014. [pdf](http://arxiv.org/abs/1403.0968) 


CHANGES
=====================================================================
NEW IN VERSION 19
=====================================================================
- (Feature) Ports of XSBench in a variety of accelerator oriented
  programming models have been added to the repo. The traditional
  OpenMP threading model of XSBench is now located in the 
  "XSBench/src/openmp-threading" directory. The other programming
  models are:

    - OpenMP target offloading, for use with OpenMP 4.5+ compilers
    that can target a variety of acecelerator architectures.

    - CUDA for use on NVIDIA GPUs.

    - OpenCL for use on CPU and accelerators. Note that this port
      may need to be heavily re-optimized if running on a non-GPU
      architecture.

  Note that the accelerator-oriented models only implement the event
  based model of parallelism, as this model is expected to be 
  superior to the history-based model for accelerator architectures.
  As such, running with the accelerator models will require the
  "-m event" flag to be used.

- (Feature) Verification mode is now default and required, so is no
  longer an option in the Makefile. A new and faster method of
  verifying results was developed and added, making it have much less
  of an impact on performance than the previous methods did. As the
  new method generates a different hash than the old one, the expected
  has values have changed in v19.

- (Feature) Changed the binary file read/write capability in XSBench
  to be a command line argument rather than requiring re-compilation
  with makefile flags. This can be set with the "-b <read, write>"
  argument.

- (Removal) Removed PAPI performance counters from code and Makefile.

- (Removal) A number of optional/outdated Makefile options were
  removed in an effort to clean up the code base and simplify things.

=====================================================================
NEW IN VERSION 18
=====================================================================
- (Feature) Added in an option to
  use an event based simulation algorithm as opposed to the default
  history based algorithm. This mode can be enabled with the "-m
  event" command line option. The central difference between the
  default history based algorithm and the event based algorithm is
  the dependence or independence of macroscopic cross section
  lookups. In the default history based method, the simulation
  parallelism is expressed over a large number of independent
  particle "histories", with each particle history requiring (by
  default) 34 sequential and dependent macroscopic cross section
  lookups to be executed. In the event based method, all macroscopic
  cross sections are pooled together in a way that allows them to
  be executed independently in any order. On CPU architectures,
  both algorithms run in approximately the same speed, though in
  full Monte Carlo applications there are some added costs associated
  with the event based method (namely, the need to frequently store
  and load particle states) that are not represented in XSBench.
  However, the model event based algorithm in XSBench is useful for
  examining how the dependence or independence of the loop over
  macroscopic cross sections may affect performance on different
  computer architectures.

  The event based mode in XSBench is very similar to the algorithm
  expressed prior to v16 and before, so performance results collected
  with v16 and previous should be comparable to performance results
  collected with the event mode in v18. On most CPU architectures
  and with most compilers, the history and event based methods run
  in approximately the same time, with some CPUs showing a small
  (less than 20%) speedup with the event based method.

- Removed the "benchmark" feature that tested the code on different
  numbers of threads. This was removed as it can easily be implemented
  in shell script form, and unnecessarily complicated the code.

- With the addition of the different event and history based options,
  the main function of the code has been broken down into some
  smaller functions. Notably, the main parallel simulation phase
  has been moved to a history and event based function in the
  "Simulation.c" file.

- Slight change to some integer castings in the verification mode
  to get consitent results between gnu and intel compilers. This
  changed the correct verification checksums, which are automatically
  checked against.

=====================================================================
NEW IN VERSION 17
=====================================================================
- In previous versions of XSBench, it was assumed that the method
  of parallelism in the app would not be altered, as the app was
  written to respect some loop dependencies of fully featured MC
  applications. However, to foster more experimentation with
  different methods of parallelizing the XS lookup kernel, we have
  decided to add an extra loop (over particles) into XSBench so that
  loop dependencies can be explicit in XSBench. This should make it
  easier for people to optimize or port XSBench without worry that
  they are breaking any implicit loop dependencies. All loop 
  dependencies should now be apparent without requiring any knowledge
  of full MC applications.

  From a performance standpoint, the change does not really affect
  anything regarding default performance, so performance results for
  CPUs (run without modification of the source code) should be
  virtually identical between v16 and v17.

  XSBench now has its default OpenMP thread level parallelism
  expressed over independent particle histories. Each particle
  then executes 34 macro_xs lookups, which are dependent, meaning
  that this loop cannot be parallelized as each iteration is
  dependent on results from the previous iteration. For each
  macro_xs, there is an independent inner loop over micro_xs lookups,

src/io.c
#include "XSbench_header.h"

#ifdef MPI
#include<mpi.h>
#endif

// Prints program logo
void logo(int version)
{
	border_print();
	printf(
	"                   __   __ ___________                 _                        \n"
	"                   \\ \\ / //  ___| ___ \\               | |                       \n"
	"                    \\ V / \\ `--.| |_/ / ___ _ __   ___| |__                     \n"
	"                    /   \\  `--. \\ ___ \\/ _ \\ '_ \\ / __| '_ \\                    \n"
	"                   / /^\\ \\/\\__/ / |_/ /  __/ | | | (__| | | |                   \n"
	"                   \\/   \\/\\____/\\____/ \\___|_| |_|\\___|_| |_|                   \n\n"
    );
	border_print();
	center_print("Developed at Argonne National Laboratory", 79);
	char v[100];
	sprintf(v, "Version: %d", version);
	center_print(v, 79);
	border_print();
}

// Prints Section titles in center of 80 char terminal
void center_print(const char *s, int width)
{
	int length = strlen(s);
	int i;
	for (i=0; i<=(width-length)/2; i++) {
		fputs(" ", stdout);
	}
	fputs(s, stdout);
	fputs("\n", stdout);
}

void print_results( Inputs in, int mype, double runtime, int nprocs,
	unsigned long long vhash )
{
	// Calculate Lookups per sec
	int lookups = 0;
	if( in.simulation_method == HISTORY_BASED )
		lookups = in.lookups * in.particles;
	else if( in.simulation_method == EVENT_BASED )
		lookups = in.lookups;
	int lookups_per_sec = (int) ((double) lookups / runtime);
	
	// If running in MPI, reduce timing statistics and calculate average
	#ifdef MPI
	int total_lookups = 0;
	MPI_Barrier(MPI_COMM_WORLD);
	MPI_Reduce(&lookups_per_sec, &total_lookups, 1, MPI_INT,
	           MPI_SUM, 0, MPI_COMM_WORLD);
	#endif
	
	// Print output
	if( mype == 0 )
	{
		border_print();
		center_print("RESULTS", 79);
		border_print();

		// Print the results
		printf("Threads:     %d\n", in.nthreads);
		#ifdef MPI
		printf("MPI ranks:   %d\n", nprocs);
		#endif
		#ifdef MPI
		printf("Total Lookups/s:            ");
		fancy_int(total_lookups);
		printf("Avg Lookups/s per MPI rank: ");
		fancy_int(total_lookups / nprocs);
		#else
		printf("Runtime:     %.3lf seconds\n", runtime);
		printf("Lookups:     "); fancy_int(lookups);
		printf("Lookups/s:   ");
		fancy_int(lookups_per_sec);
		#endif

		unsigned long long large = 0;
		unsigned long long small = 0; 
		if( in.simulation_method == EVENT_BASED )
		{
			small = 387203;
			large = 666382;
		}
		else if( in.simulation_method == HISTORY_BASED )
		{
			large = 658515;
			small = 385793; 
		}
		if( strcmp(in.HM, "large") == 0 )
		{
			if( vhash == large )
				printf("Verification checksum: %llu (Valid)\n", vhash);
			else
				printf("Verification checksum: %llu (WARNING - INAVALID CHECKSUM!)\n", vhash);
		}
		else if( strcmp(in.HM, "small") == 0 )
		{
			if( vhash == small )
				printf("Verification checksum: %llu (Valid)\n", vhash);
			else
				printf("Verification checksum: %llu (WARNING - INAVALID CHECKSUM!)\n", vhash);
		}
		border_print();
	}
}

void print_inputs(Inputs in, int nprocs, int version )
{
	// Calculate Estimate of Memory Usage
	int mem_tot = estimate_mem_usage( in );
	logo(version);
	center_print("INPUT SUMMARY", 79);
	border_print();
	if( in.simulation_method == EVENT_BASED )
		printf("Simulation Method:            Event Based\n");
	else
		printf("Simulation Method:            History Based\n");
	if( in.grid_type == NUCLIDE )
		printf("Grid Type:                    Nuclide Grid\n");
	else if( in.grid_type == UNIONIZED )
		printf("Grid Type:                    Unionized Grid\n");
	else
		printf("Grid Type:                    Hash\n");

	printf("Materials:                    %d\n", 12);
	printf("H-M Benchmark Size:           %s\n", in.HM);
	printf("Total Nuclides:               %ld\n", in.n_isotopes);
	printf("Gridpoints (per Nuclide):     ");
	fancy_int(in.n_gridpoints);
	if( in.grid_type == HASH )
	{
		printf("Hash Bins:                    ");
		fancy_int(in.hash_bins);
	}
	if( in.grid_type == UNIONIZED )
	{
		printf("Unionized Energy Gridpoints:  ");
		fancy_int(in.n_isotopes*in.n_gridpoints);
	}
	if( in.simulation_method == HISTORY_BASED )
	{
		printf("Particle Histories:           "); fancy_int(in.particles);
		printf("XS Lookups per Particle:      "); fancy_int(in.lookups);
	}
	printf("Total XS Lookups:             "); fancy_int(in.lookups);
	#ifdef MPI
	printf("MPI Ranks:                    %d\n", nprocs);
	printf("OMP Threads per MPI Rank:     %d\n", in.nthreads);
	printf("Mem Usage per MPI Rank (MB):  "); fancy_int(mem_tot);
	#else
	printf("Threads:                      %d\n", in.nthreads);
	printf("Est. Memory Usage (MB):       "); fancy_int(mem_tot);
	#endif
	printf("Binary File Mode:             ");
	if( in.binary_mode == NONE )
		printf("Off\n");
	else if( in.binary_mode == READ)
		printf("Read\n");
	else
		printf("Write\n");
	border_print();
	center_print("INITIALIZATION - DO NOT PROFILE", 79);
	border_print();
}

void border_print(void)
{
	printf(
	"==================================================================="
	"=============\n");
}

// Prints comma separated integers - for ease of reading
void fancy_int( long a )
{
    if( a < 1000 )
        printf("%ld\n",a);

    else if( a >= 1000 && a < 1000000 )
        printf("%ld,%03ld\n", a / 1000, a % 1000);

    else if( a >= 1000000 && a < 1000000000 )
        printf("%ld,%03ld,%03ld\n",a / 1000000,(a % 1000000) / 1000,a % 1000 );

    else if( a >= 1000000000 )
        printf("%ld,%03ld,%03ld,%03ld\n",
               a / 1000000000,
               (a % 1000000000) / 1000000,
               (a % 1000000) / 1000,
               a % 1000 );
    else
        printf("%ld\n",a);
}

void print_CLI_error(void)
{
	printf("Usage: ./XSBench <options>\n");
	printf("Options include:\n");
	printf("  -m <simulation method>   Simulation method (history, event)\n");
	printf("  -t <threads>             Number of OpenMP threads to run\n");
	printf("  -s <size>                Size of H-M Benchmark to run (small, large, XL, XXL)\n");
	printf("  -g <gridpoints>          Number of gridpoints per nuclide (overrides -s defaults)\n");
	printf("  -G <grid type>           Grid search type (unionized, nuclide, hash). Defaults to unionized.\n");
	printf("  -p <particles>           Number of particle histories\n");
	printf("  -l <lookups>             History Based: Number of Cross-section (XS) lookups per particle. Event Based: Total number of XS lookups.\n");
	printf("  -h <hash bins>           Number of hash bins (only relevant when used with \"-G hash\")\n");
	printf("  -b <binary mode>         Read or write all data structures to file. If reading, this will skip initialization phase. (read, write)\n");
	printf("Default is equivalent to: -m history -s large -l 34 -p 500000 -G unionized\n");
	printf("See readme for full description of default run values\n");
	exit(4);
}

Inputs read_CLI( int argc, char * argv[] )
{
	Inputs input;

	// defaults to the history based simulation method
	input.simulation_method = HISTORY_BASED;
	
	// defaults to max threads on the system	
	input.nthreads = omp_get_num_procs();
	
	// defaults to 355 (corresponding to H-M Large benchmark)
	input.n_isotopes = 355;
	
	// defaults to 11303 (corresponding to H-M Large benchmark)
	input.n_gridpoints = 11303;

	// defaults to 500,000
	input.particles = 500000;
	
	// defaults to 34
	input.lookups = 34;
	
	// default to unionized grid
	input.grid_type = UNIONIZED;

	// default to unionized grid
	input.hash_bins = 10000;

	// default to no binary read/write
	input.binary_mode = NONE;
	
	// defaults to H-M Large benchmark
	input.HM = (char *) malloc( 6 * sizeof(char) );
	input.HM[0] = 'l' ; 
	input.HM[1] = 'a' ; 
	input.HM[2] = 'r' ; 
	input.HM[3] = 'g' ; 
	input.HM[4] = 'e' ; 
	input.HM[5] = '\0';
	
	// Check if user sets these
	int user_g = 0;

	int default_lookups = 1;
	int default_particles = 1;
	
	// Collect Raw Input
	for( int i = 1; i < argc; i++ )
	{
		char * arg = argv[i];

		// nthreads (-t)
		if( strcmp(arg, "-t") == 0 )
		{
			if( ++i < argc )
				input.nthreads = atoi(argv[i]);
			else
				print_CLI_error();
		}
		// n_gridpoints (-g)
		else if( strcmp(arg, "-g") == 0 )
		{	
			if( ++i < argc )
			{
				user_g = 1;
				input.n_gridpoints = atol(argv[i]);
			}
			else
				print_CLI_error();
		}
		// Simulation Method (-m)
		else if( strcmp(arg, "-m") == 0 )
		{
			char * sim_type;
			if( ++i < argc )
				sim_type = argv[i];
			else
				print_CLI_error();

			if( strcmp(sim_type, "history") == 0 )
				input.simulation_method = HISTORY_BASED;
			else if( strcmp(sim_type, "event") == 0 )
			{
				input.simulation_method = EVENT_BASED;
				// Also resets default # of lookups
				if( default_lookups && default_particles )
				{
					input.lookups =  input.lookups * input.particles;
					input.particles = 0;
				}
			}
			else
				print_CLI_error();
		}
		// lookups (-l)
		else if( strcmp(arg, "-l") == 0 )
		{
			if( ++i < argc )
			{
				input.lookups = atoi(argv[i]);
				default_lookups = 0;
			}
			else
				print_CLI_error();
		}
		// hash bins (-h)
		else if( strcmp(arg, "-h") == 0 )
		{
			if( ++i < argc )
				input.hash_bins = atoi(argv[i]);
			else
				print_CLI_error();
		}
		// particles (-p)
		else if( strcmp(arg, "-p") == 0 )
		{
			if( ++i < argc )
			{
				input.particles = atoi(argv[i]);
				default_particles = 0;
			}
			else
				print_CLI_error();
		}
		// HM (-s)
		else if( strcmp(arg, "-s") == 0 )
		{	
			if( ++i < argc )
				input.HM = argv[i];
			else
				print_CLI_error();
		}
		// grid type (-G)
		else if( strcmp(arg, "-G") == 0 )
		{
			char * grid_type;
			if( ++i < argc )
				grid_type = argv[i];
			else
				print_CLI_error();

			if( strcmp(grid_type, "unionized") == 0 )
				input.grid_type = UNIONIZED;
			else if( strcmp(grid_type, "nuclide") == 0 )
				input.grid_type = NUCLIDE;
			else if( strcmp(grid_type, "hash") == 0 )
				input.grid_type = HASH;
			else
				print_CLI_error();
		}
		// binary mode (-b)
		else if( strcmp(arg, "-b") == 0 )
		{
			char * binary_mode;
			if( ++i < argc )
				binary_mode = argv[i];
			else
				print_CLI_error();

			if( strcmp(binary_mode, "read") == 0 )
				input.binary_mode = READ;
			else if( strcmp(binary_mode, "write") == 0 )
				input.binary_mode = WRITE;
			else
				print_CLI_error();
		}
		else
			print_CLI_error();
	}

	// Validate Input
	
	// Validate nthreads
	if( input.nthreads < 1 )
		print_CLI_error();
	
	// Validate n_isotopes
	if( input.n_isotopes < 1 )
		print_CLI_error();
	
	// Validate n_gridpoints
	if( input.n_gridpoints < 1 )
		print_CLI_error();

	// Validate lookups
	if( input.lookups < 1 )
		print_CLI_error();

	// Validate Hash Bins 
	if( input.hash_bins < 1 )
		print_CLI_error();
	
	// Validate HM size
	if( strcasecmp(input.HM, "small") != 0 &&
		strcasecmp(input.HM, "large") != 0 &&
		strcasecmp(input.HM, "XL") != 0 &&
		strcasecmp(input.HM, "XXL") != 0 )
		print_CLI_error();
	
	// Set HM size specific parameters
	// (defaults to large)
	if( strcasecmp(input.HM, "small") == 0 )
		input.n_isotopes = 68;
	else if( strcasecmp(input.HM, "XL") == 0 && user_g == 0 )
		input.n_gridpoints = 238847; // sized to make 120 GB XS data
	else if( strcasecmp(input.HM, "XXL") == 0 && user_g == 0 )
		input.n_gridpoints = 238847 * 2.1; // 252 GB XS data

	// Return input struct
	return input;
}

void binary_write( Inputs in, SimulationData SD )
{
	char * fname = "XS_data.dat";
	printf("Writing all data structures to binary file %s...\n", fname);
	FILE * fp = fopen(fname, "w");

	// Write SimulationData Object. Include pointers, even though we won't be using them.
	fwrite(&SD, sizeof(SimulationData), 1, fp);

	// Write heap arrays in SimulationData Object
	fwrite(SD.num_nucs,       sizeof(int), SD.length_num_nucs, fp);
	fwrite(SD.concs,          sizeof(double), SD.length_concs, fp);
	fwrite(SD.mats,           sizeof(int), SD.length_mats, fp);
	fwrite(SD.nuclide_grid,   sizeof(NuclideGridPoint), SD.length_nuclide_grid, fp); 
	fwrite(SD.index_grid, sizeof(int), SD.length_index_grid, fp);
	fwrite(SD.unionized_energy_array, sizeof(double), SD.length_unionized_energy_array, fp);

	fclose(fp);
}

SimulationData binary_read( Inputs in )
{
	SimulationData SD;
	
	char * fname = "XS_data.dat";
	printf("Reading all data structures from binary file %s...\n", fname);

	FILE * fp = fopen(fname, "r");
	assert(fp != NULL);

	// Read SimulationData Object. Include pointers, even though we won't be using them.
	fread(&SD, sizeof(SimulationData), 1, fp);

	// Allocate space for arrays on heap
	SD.num_nucs = (int *) malloc(SD.length_num_nucs * sizeof(int));
	SD.concs = (double *) malloc(SD.length_concs * sizeof(double));
	SD.mats = (int *) malloc(SD.length_mats * sizeof(int));
	SD.nuclide_grid = (NuclideGridPoint *) malloc(SD.length_nuclide_grid * sizeof(NuclideGridPoint));
	SD.index_grid = (int *) malloc( SD.length_index_grid * sizeof(int));
	SD.unionized_energy_array = (double *) malloc( SD.length_unionized_energy_array * sizeof(double));

	// Read heap arrays into SimulationData Object
	fread(SD.num_nucs,       sizeof(int), SD.length_num_nucs, fp);
	fread(SD.concs,          sizeof(double), SD.length_concs, fp);
	fread(SD.mats,           sizeof(int), SD.length_mats, fp);
	fread(SD.nuclide_grid,   sizeof(NuclideGridPoint), SD.length_nuclide_grid, fp); 
	fread(SD.index_grid, sizeof(int), SD.length_index_grid, fp);
	fread(SD.unionized_energy_array, sizeof(double), SD.length_unionized_energy_array, fp);

	fclose(fp);

	return SD;
}
